pkg/jx/cmd/prompt_test.go

```diff
@@ -11,8 +12,7 @@ func TestNewJXCommand(t *testing.T) {
 	// TODO use mock stdio and expect some output ending in '>'
 	f := cmd.NewFactory()
 	f.SetOffline(true)
-	cmd := cmd.NewJXCommand(f, os.Stdin, os.Stdout, os.Stderr)
-	cmd.SetArgs([]string{"prompt"})
+	cmd := cmd.NewCmdPrompt(f, os.Stdin, os.Stdout, os.Stderr)
```

Comment:
Did you actually try running this with the main fixes reverted one by one? It doubt it will still fail as expected. The reason I wrote the test the way I did is because the chain of events starting from `NewJXCommand` were what included network connections, and `NewJXCommand` seemed to be the highest up I could go in the call chain without mocking out the native `argv` or something like that. I am _not_ testing the functionality of `jx prompt` per se; this is just an example of a command that is not expected to make any network connection at any point in its lifecycle.

Reply:
"Did you actually try running this with the main fixes reverted one by one?"

Not 100% sure what you mean (if you mean did I run the test in each of your commits, then no, I just ran `HEAD`), but I ran your test in the branch in your PR and it failed.

So, I think I misunderstood the issue you raised :-) What do you want to happen without network access? My understanding from the discussion was that `jx prompt` shouldn't require network access (which I agree with). It would help me to understand what commands you want to run without network access so I can try to redesign accordingly.

Reply:
> I ran your test in the branch in your PR

No, I meant for example:

```sh
git checkout 09191c288924899ef7082f0e328a2524fb8eb278 # HEAD of #2414
git checkout master -- pkg/jx/cmd/add_app.go # one example
(cd pkg/jx/cmd; go test -run TestNewJXCommand)
```

resulting in

```
--- FAIL: TestNewJXCommand (0.00s)
panic: not supposed to be making a network connection [recovered]
	panic: not supposed to be making a network connection

goroutine 25 [running]:
testing.tRunner.func1(0xc00049bf00)
	/snap/go/3095/src/testing/testing.go:792 +0x387
panic(0x23d07a0, 0x2decf20)
	/snap/go/3095/src/runtime/panic.go:513 +0x1b9
github.com/jenkins-x/jx/pkg/jx/cmd.createKubeConfig(0x2a2f901, 0xa)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/factory.go:456 +0x197
github.com/jenkins-x/jx/pkg/jx/cmd.(*factory).CreateKubeConfig(0xc0004b5180, 0x0, 0x0, 0x0)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/factory.go:480 +0x21b
github.com/jenkins-x/jx/pkg/jx/cmd.(*factory).CreateJXClient(0xc0004b5180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x8)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/factory.go:356 +0x2f
github.com/jenkins-x/jx/pkg/jx/cmd.(*CommonOptions).JXClient(0xc000211600, 0x0, 0x0, 0x0, 0x203000, 0x203000, 0x0)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/common.go:196 +0x8c
github.com/jenkins-x/jx/pkg/jx/cmd.(*CommonOptions).JXClientAndDevNamespace(0xc000211600, 0xc0005b7390, 0x10, 0x10, 0x26027c0, 0x203000, 0x0)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/common.go:241 +0x170
github.com/jenkins-x/jx/pkg/jx/cmd.(*CommonOptions).GetDevEnv(0xc000211600, 0x0, 0x10)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/common_app.go:19 +0x2f
github.com/jenkins-x/jx/pkg/jx/cmd.(*AddAppOptions).addFlags(0xc000211600, 0xc0009af180, 0x2a24037, 0x2, 0x0, 0x0, 0x0, 0x0)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/add_app.go:78 +0x31
github.com/jenkins-x/jx/pkg/jx/cmd.NewCmdAddApp(0x2e5a4e0, 0xc0004b5180, 0x2e0e3e0, 0xc00000e010, 0x2e0e420, 0xc00000e018, 0x2dfff00, 0xc00000e020, 0x280)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/add_app.go:72 +0x188
github.com/jenkins-x/jx/pkg/jx/cmd.NewCmdAdd(0x2e5a4e0, 0xc0004b5180, 0x2e0e3e0, 0xc00000e010, 0x2e0e420, 0xc00000e018, 0x2dfff00, 0xc00000e020, 0x0)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/add.go:55 +0x1eb
github.com/jenkins-x/jx/pkg/jx/cmd.NewJXCommand(0x2e5a4e0, 0xc0004b5180, 0x2e0e3e0, 0xc00000e010, 0x2e0e420, 0xc00000e018, 0x2dfff00, 0xc00000e020, 0x5c2f8d53)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/cmd.go:63 +0x108
github.com/jenkins-x/jx/pkg/jx/cmd_test.TestNewJXCommand(0xc00049bf00)
	/src/gh/jenkins-x/jx/pkg/jx/cmd/cmd_test.go:14 +0xea
testing.tRunner(0xc00049bf00, 0x2b51738)
	/snap/go/3095/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/snap/go/3095/src/testing/testing.go:878 +0x35c
exit status 2
FAIL	github.com/jenkins-x/jx/pkg/jx/cmd	0.067s
```

as expected, pinpointing `addFlags` as the problem.

> It would help me to understand what commands you want to run without network access

Well, _any_ command that does not specifically request network access. The problem was that _every_ command (including ones like `prompt` which ought to pay attention solely to local files) was making a network connection by way of plumbing invoked directly or indirectly from `NewJXCommand` while merely setting up the list of _available_ commands. In the example above, this code was being run even though you were not running `jx add app`.

Reply:
> I ran your test in the branch in your PR and it failed

Failed how? Passes for me locally (also after `git merge master`) and apparently in CI.

Reply:
> ...as expected, pinpointing addFlags as the problem.

Ah, I see what you mean. No, I didn't do that as I know that the `addFlags` thing was what was causing the network access.

> Failed how? Passes for me locally (also after git merge master) and apparently in CI.

Will rerun to check.

> Well, any command that does not specifically request network access.

`jx help` (which is what `jx` ends up invoking) does *require* network access, as it needs to check if there are any plugins to load. This is why I isolated the test to a command that doesn't (shouldn't) require network access.

Reply:
> `jx help` (which is what `jx` ends up invoking) does _require_ network access, as it needs to check if there are any plugins to load.

Right, that is OK, and that is why I was not using `jx help` as an example of a network-free command in the test.

> I isolated the test to a command that doesn't (shouldn't) require network access

The test was already isolated to the `jx prompt` command. But the change to call `NewCmdPrompt` directly means you are bypassing all the problematic code I was fixing! With your change, the test passes no matter how broken everything else is.

Reply:
For example:

```sh
git checkout 09191c288924899ef7082f0e328a2524fb8eb278
git checkout master -- pkg/jx/cmd/add_app.go
(cd pkg/jx/cmd; go test -run TestNewJXCommand) # fails, good
vi pkg/jx/cmd/cmd_test.go
# -cmd := cmd.NewJXCommand(f, os.Stdin, os.Stdout, os.Stderr)
# -cmd.SetArgs([]string{"prompt"})
# +cmd := cmd.NewCmdPrompt(f, os.Stdin, os.Stdout, os.Stderr)
(cd pkg/jx/cmd; go test -run TestNewJXCommand) # passes when it should not!
```

Reply:
This is the failure on your branch (merged with HEAD of master):

```GOROOT=/usr/local/Cellar/go/1.11.4/libexec #gosetup
GOPATH=/Users/pmuir/go #gosetup
/usr/local/Cellar/go/1.11.4/libexec/bin/go test -c -o /private/var/folders/qr/h482727j4dvc7qphqfkvws1r0000gn/T/___cmd_test_go github.com/jenkins-x/jx/pkg/jx/cmd #gosetup
/usr/local/Cellar/go/1.11.4/libexec/bin/go tool test2json -t /private/var/folders/qr/h482727j4dvc7qphqfkvws1r0000gn/T/___cmd_test_go -test.v -test.run ^TestNewJXCommand$ #gosetup
=== RUN   TestNewJXCommand
--- FAIL: TestNewJXCommand (0.00s)
panic: not supposed to be making a network connection [recovered]
	panic: not supposed to be making a network connection

goroutine 10 [running]:
testing.tRunner.func1(0xc0006a9600)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:792 +0x387
panic(0x5fcc720, 0x69eb9b0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/panic.go:513 +0x1b9
github.com/jenkins-x/jx/pkg/jx/cmd.createKubeConfig(0x662e101, 0xa)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/factory.go:456 +0x197
github.com/jenkins-x/jx/pkg/jx/cmd.(*factory).CreateKubeConfig(0xc000a28910, 0x0, 0x0, 0x0)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/factory.go:480 +0x21b
github.com/jenkins-x/jx/pkg/jx/cmd.(*factory).CreateApiExtensionsClient(0xc000a28910, 0x0, 0x0, 0x0, 0x0)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/factory.go:407 +0x2f
github.com/jenkins-x/jx/pkg/jx/cmd.(*CommonOptions).CreateApiExtensionsClient(0xc0007883c0, 0x0, 0x0, 0x0, 0x0)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/common.go:158 +0x6b
github.com/jenkins-x/jx/pkg/jx/cmd.(*CommonOptions).isManagedPluginsEnabled(0xc0007883c0, 0x0)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/common_plugins.go:24 +0x40
github.com/jenkins-x/jx/pkg/jx/cmd.(*CommonOptions).getPluginCommandGroups(0xc0007883c0, 0x69fc2c0, 0xc00079a340, 0xc0006e3970, 0x41f93e6, 0xc00028a780, 0xc00079a300, 0x1, 0x1)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/common_plugins.go:51 +0x111
github.com/jenkins-x/jx/pkg/jx/cmd.NewJXCommand.func1(0xc00028a780, 0xc0000d40d0, 0x3, 0x3)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/cmd.go:200 +0xa1
github.com/jenkins-x/jx/pkg/jx/cmd.NewJXCommand(0x6a595a0, 0xc000a28910, 0x6a0cee0, 0xc0000de000, 0x6a0cf20, 0xc0000de008, 0x69fe9e0, 0xc0000de010, 0x5c3333f3)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/cmd.go:226 +0x2bce
github.com/jenkins-x/jx/pkg/jx/cmd_test.TestNewJXCommand(0xc0006a9600)
	/Users/pmuir/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/cmd_test.go:14 +0xea
testing.tRunner(0xc0006a9600, 0x674fe50)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c

Process finished with exit code 1
```

Reply:
This is due to the way we are directly reading the `os.Args` so we end up with the test tool args getting passed in. I'll move this around to make it possible to directly pass the correct arg in.

Reply:
Huh, not sure what those commands you are running are. I could not reproduce using a simple

```sh
(cd pkg/jx/cmd; go test -run TestNewJXCommand)
```

Reply:
Right click in Goland and select Run for the test :-)